### PR TITLE
blopost about test design

### PR DESCRIPTION
**Pull request by pvanhees**

### CODE REVIEW

File: _posts/2019-07-01-test-design.md [hunk 1]
```diff
@@ -0,0 +1,294 @@
...
+        return this.customerId;
+    }
+}
+```
+
+```java
+public class OrderTest {
+    
+    @Test
+    public void createOrder_happyPath() {
```
Comment (yannickdeturck) [inline _posts/2019-07-01-test-design.md]: I personally like to see test names as sentences that describe what is being tested exactly.
While in Java we are stuck with the regular method naming restrictions we can utilise underscores to write our test name as a sentence.

```suggestion
    public void given_an_active_customer_then_order_creation_should_be_successful() {
```

In Kotlin for example this is slightly more convenient as you can write actual sentences by wrapping the method in backticks.

File: _posts/2019-07-01-test-design.md [hunk 2]
```diff
@@ -0,0 +1,294 @@
...
+        when(customer.getId()).thenReturn(customerId);
+        when(customer.isActive()).thenReturn(true);
+        
+        final Order order = new Order(customer);
+        
+        assertThat(order.getCustomerId()).isEqualTo(customerId);
+    }
+    
+    @Test
+    public void createOrder_customerInactive() {
```
Comment (yannickdeturck) [inline _posts/2019-07-01-test-design.md]: We can also do the same here:
```suggestion
    public void given_an_inactive_customer_then_order_creation_should_result_in_an_illegal_argument_exception() {
```

File: _posts/2019-07-01-test-design.md [hunk 3]
```diff
@@ -0,0 +1,294 @@
...
+    }
+    
+    public UUID getCustomerId() {
+        return this.customerId;
+    }
+}
+```
+
+However when we tried to run all the tests, most of them didn't compile anymore, which makes sense because we changed the contract.
+Now we could have made it easier for ourselves by using some Intellij refactoring tools, but nevertheless, it's absurd that so many tests could break by just changing the way we construct our aggregates.
```
Comment (yannickdeturck) [inline _posts/2019-07-01-test-design.md]:
```suggestion
Now we could have made it easier for ourselves by using some IntelliJ refactoring tools, but nevertheless, it's absurd that so many tests could break by just changing the way we construct our aggregates.
```

File: _posts/2019-07-01-test-design.md [hunk 4]
```diff
@@ -0,0 +1,294 @@
...
+Not having learned from our previous mistakes and being under time pressure, we decided to use the factory to create instances in all our tests. 
+For the factory we needed some other services, repositories etc. which we all mocked.
+This was a huge amount of work because of all the mocking we had to do just so we could create a consistent aggregate.
+And we had to do this, again, in every test that needs an aggregate.
+
+After everything worked again, we were happy that the pile of work was done and we could move on with other things.
+
+In the weeks that followed, however, we started to notice that every time we changed the logic of the factory, we needed to change all the tests again because we had to add some extra mocks, data, etc. in all the tests.
+
+After a few of these iterations where we had to spend too many time fixing tests, we were fed up and decided (way too late of course) to free up some time for a more structural solution.
```
Comment (yannickdeturck) [inline _posts/2019-07-01-test-design.md]:
```suggestion
After a few of these iterations where we had to spend too much time fixing tests, we were fed up and decided (way too late of course) to free up some time for a more structural solution.
```

File: _posts/2019-07-01-test-design.md [hunk 5]
```diff
@@ -0,0 +1,294 @@
...
+We got some inspiration from a [blogpost](https://blog.cleancoder.com/uncle-bob/2017/03/03/TDD-Harms-Architecture.html){:target="_blank" rel="noopener noreferrer"} from Uncle Bob about his opinion on the statement that *TDD harms architecture*.
+One of the things he mentions in his post is that we shouldn't make the mistake of coupling every test to the implementation of our production code.
+Instead it would be better to put some sort of API in between our tests and the production code.
+
+## Implementing the API 
+We didn't take Uncle Bob's solution too literally and gave our own twist to it.
+
+For the specific problem of creating aggregate instances we decided to create a class that acts as a scenario builder.
+In this CreateOrderScenario we have a static factory method that will create a scenario that returns a valid Order when executed.
+This means that when you need an order that is consistent and it doesn't matter for your test which data is in order, you can just use the default scenario when it's executed.
```
Comment (yannickdeturck) [inline _posts/2019-07-01-test-design.md]:
```suggestion
This means that when you need an order that is consistent and it doesn't matter for your test which data is in the order, you can just use the default scenario when it's executed.
```

File: _posts/2019-07-01-test-design.md [hunk 6]
```diff
@@ -0,0 +1,294 @@
...
+
+You could also create other default scenarios.
+For example an order with an invalid customer, or with specific data that triggers a certain flow in the order process.
+
+This is very convenient for most tests.
+However, in some tests we want to influence how the Order is constructed, so we can test some custom cases other than a default scenario, specific for certain tests.
+We implemented this by adding some methods to our scenario class in that allow the scenario to be modified to the test's needs.
```
Comment (yannickdeturck) [inline _posts/2019-07-01-test-design.md]:
```suggestion
We implemented this by adding some methods to our scenario class that allow the scenario to be modified to the test's needs.
```

File: _posts/2019-07-01-test-design.md [hunk 7]
```diff
@@ -0,0 +1,294 @@
...
+        return orderFactory.createOrder(createOrderData);
+    }
+}
+```
+
+In the following example we created two tests that verify that the construction of an order works correctly.
+In the first test, we use the default scenario without modifying anything, meaning, we test the happy path and verify that all data in the created order is correct.
+
+In the second test we verify that if we try to create an order for a customer that doesn't exist, we get a validation exception.
+We do this by creating a default scenario, then modifying the input data to use a customerId defined in the test, and then overriding the behaviour of the CustomerRepository mock.
```
Comment (yannickdeturck) [inline _posts/2019-07-01-test-design.md]:
```suggestion
We do this by creating a default scenario, then modifying the input data to use a `customerId` defined in the test, and then overriding the behaviour of the `CustomerRepository` mock.
```

File: _posts/2019-07-01-test-design.md [hunk 8]
```diff
@@ -0,0 +1,294 @@
...
+        assertThat(order.getCustomerId()).isEqualTo(CreateOrderScenario.customerId);
+    }
+    
+    @Test
+    void createOrder_customerInactive() {
+        final UUID customerId = UUID.randomUUID();
+        final CreateOrderScenario scenario = CreateOrderScenario
+            .defaultScenario()
+            .modifyCreateOrderData(orderData -> orderData.setCustomerId(customerId))
+            .overrideCustomerRepository(repository ->
```
Comment (yannickdeturck) [inline _posts/2019-07-01-test-design.md]: In this case overriding the mocking shouldn't be necessary as the default scenario specifically mocks the `findById` call with a specific `customerId`. So all calls with a different `customerId` will result in a `null` by default.
But I understand the example you want to make. :)

File: _posts/2019-07-01-test-design.md [hunk 9]
```diff
@@ -0,0 +1,294 @@
...
+* the use of a factory to create orders,
+* a validator class, used by the factory to validate the input for creating an order,
+* and how the constructor of the Order aggregate should be called. 
+
+With this example, it's easy to see that our new tests are much less likely to break than our original **design**. 
+There is a clean layer between the implementation/design of our application, and the tests.
+This lower coupling makes it easier to refactor the application, and implement new features at a higher pace.
+
+Also notice that we didn't create a test class that maps one-to-one to a production code class.
+Rather than testing our Order object, or our OrderFactory, or our OrderValidator, we test the creation of an aggregate instance.
```
Comment (yannickdeturck) [inline _posts/2019-07-01-test-design.md]:
```suggestion
Rather than testing our `Order` object, or our `OrderFactory`, or our `OrderValidator`, we test the creation of an aggregate instance. 
```

File: _posts/2019-07-01-test-design.md [hunk 10]
```diff
@@ -0,0 +1,294 @@
...
+    }
+}
+```
+
+## Advantages of using an API
+
+The advantage of this design is that our tests are not aware of:
+* the use of a factory to create orders,
+* a validator class, used by the factory to validate the input for creating an order,
+* and how the constructor of the Order aggregate should be called.
```
Comment (yannickdeturck) [inline _posts/2019-07-01-test-design.md]:
```suggestion
* and how the constructor of the `Order` aggregate should be called. 
```

File: _posts/2019-07-01-test-design.md [hunk 11]
```diff
@@ -0,0 +1,294 @@
...
+
+With this example, it's easy to see that our new tests are much less likely to break than our original **design**. 
+There is a clean layer between the implementation/design of our application, and the tests.
+This lower coupling makes it easier to refactor the application, and implement new features at a higher pace.
+
+Also notice that we didn't create a test class that maps one-to-one to a production code class.
+Rather than testing our Order object, or our OrderFactory, or our OrderValidator, we test the creation of an aggregate instance. 
+We test what we expect our application to do, not what we expect our class to do.
+
+Whenever we have to change the logic of how an Order is created, we know that we have to look in the CreateOrderTest class.
```
Comment (yannickdeturck) [inline _posts/2019-07-01-test-design.md]:
```suggestion
Whenever we have to change the logic of how an `Order` is created, we know that we have to look in the `CreateOrderTest` class.
```

File: _posts/2019-07-01-test-design.md [hunk 12]
```diff
@@ -0,0 +1,294 @@
...
+We don't have to look at the OrderTest class, the OrderFactoryTest class, or the OrderValidator test class to see where we should add some tests.
```
Comment (yannickdeturck) [inline _posts/2019-07-01-test-design.md]:
```suggestion
We don't have to look at the `OrderTest` class, the `OrderFactoryTest` class, or the `OrderValidator` test class to see where we should add some tests.
```

File: _posts/2019-07-01-test-design.md [hunk 13]
```diff
@@ -0,0 +1,294 @@
...
+
+## Conclusion
```
Comment (yannickdeturck) [inline _posts/2019-07-01-test-design.md]: Good conclusion. 👍🏼

File: _posts/2019-07-01-test-design.md (hunk 8 above)
Comment (pvanhees) [reply _posts/2019-07-01-test-design.md]: fixed by letting the repository return an optional ;)

File: _posts/2019-07-01-test-design.md [hunk 14]
```diff
@@ -0,0 +1,296 @@
+---
+layout: post
+authors: [pieter_van_hees]
+title: 'Design your tests!'
+image: /img/devoxx-fr-2019/devoxx-fr.png
+tags: [Test Driven Development, Test Design, Architecture]
```
Comment (kencoenen) [inline _posts/2019-07-01-test-design.md]: We already have a `Testing` tag too, add it too.

File: _posts/2019-07-01-test-design.md [hunk 15]
```diff
@@ -0,0 +1,296 @@
...
+    public void given_an_inactive_customer_then_order_creation_should_result_in_an_illegal_argument_exception() {
+        final Customer customer = mock(Customer.class);
+        when(customer.isActive()).thenReturn(false);
+ 
+        assertThrows(IllegalArgumentException.class, () -> new Order(customer));
+    }
+}
+```
+
+Our problems began when we realised that the constructor of our aggregate was becoming too large.
```
Comment (kencoenen) [inline _posts/2019-07-01-test-design.md]: I would rephrase here that the number of parameters became too large.
When I read the second part about the complex logic in the constructor, I thought you had written the same thing twice.

File: _posts/2019-07-01-test-design.md [hunk 16]
```diff
@@ -0,0 +1,296 @@
...
+        
+        assertThrows(IllegalArgumentException.class, () -> new Order(data));
+    }
+}
+```
+
+A few months later, after adding some more features, we noticed that there was too much logic inside the constructor of our aggregate. 
+The constructor became too big and complex so we decided to use the factory pattern to create new instances of the aggregate. 
+
+As we didn't change the input parameters of the constructor, all our tests still compiled and we were happy.
```
Comment (kencoenen) [inline _posts/2019-07-01-test-design.md]: "Since we used a `CreateOrderData` class..." makes it more clear I guess.

File: _posts/2019-07-01-test-design.md [hunk 17]
```diff
@@ -0,0 +1,296 @@
...
+
+We got some inspiration from a [blogpost](https://blog.cleancoder.com/uncle-bob/2017/03/03/TDD-Harms-Architecture.html){:target="_blank" rel="noopener noreferrer"} from Uncle Bob about his opinion on the statement that *TDD harms architecture*.
+One of the things he mentions in his post is that we shouldn't make the mistake of coupling every test to the implementation of our production code.
+Instead it would be better to put some sort of API in between our tests and the production code.
+
+## Implementing the API 
+We didn't take Uncle Bob's solution too literally and gave our own twist to it.
+
+For the specific problem of creating aggregate instances we decided to create a class that acts as a scenario builder.
+In this CreateOrderScenario we have a static factory method that will create a scenario that returns a valid Order when executed.
```
Comment (kencoenen) [inline _posts/2019-07-01-test-design.md]: Add backticks to `CreateOrderScenario`
